getumbrel/umbrel
# OTA Updates

## Over-The-Air Updates

This PR adds the functionality to allow the user to update their Umbrel installation on the go, without having to manually install new releases or re-flash Umbrel OS image.

The execution flow of the update process is [documented here](https://github.com/mayankchhabra/umbrel/blob/ota-updates/bin/README.md).

### How to test

4 demo releases have been made on the `mayankchhabra/getumbrel` repo.

Umbrel `v0.1.4-beta.2` (base release)
Umbrel `v0.1.4-beta.3` (requires `>= v0.1.4-beta.2`)
Umbrel `v0.1.4-beta.4` (requires `>= v0.1.4-beta.3`)
Umbrel `v0.1.4-beta.5` (requires `>= v0.1.4-beta.3`)

You should install the base release - Umbrel `v0.1.1-beta.2`.

### Expected behavior

After installing `v0.1.4-beta.2` (base release), you should be able to install `v0.1.4-beta.3` and then `v0.1.4-beta.5`, thereby skipping `v0.1.4-beta.4`.

### Step-by-step testing instructions

> If testing on `umbrel-dev`, make sure `rsync` and `jq` are installed.
 
1. Create a new directory and install the base testing release inside it with:
 
```
wget -qO- "https://raw.githubusercontent.com/mayankchhabra/umbrel/v0.1.4-beta.2/install-box.sh" | sh
```

2. Configure the installation for regtest

```
REGTEST=1 ./configure-box.sh
```
> If testing on `umbrel-dev`, update the hostname to `umbrel-dev.local` by running `sed -i "s/umbrel.local/umbrel-dev.local/g;" docker-compose.yml`

3. Run `karen`

```
sudo ./karen
```

4. Start the services (in another tab)

```
docker-compose up -d
```

5. Finish the onboarding and reach the dashboard. Go to the settings page and verify that you're running Umbrel version `0.1.4-beta.2` as seen on the "system" card.

6. Meanwhile, you should also see an alert notifying about a new Umbrel release available. It should be `Umbrel v0.1.4-beta.3`. Click "install" to install the update.

7. After the update is complete, verify that the Umbrel version on the settings page is `0.1.4-beta.3`.

8. Meanwhile, you'll get another alert notifying about a new Umbrel release available. It should be `Umbrel v0.1.4-beta.5`. Click "install" to install the update.

9. After the update is complete, verify that the Umbrel version on the settings page is `0.1.4-beta.5`.

---

**Further improvements:**

1. Creating a `rollback.sh` script in `bin/update/` that rollbacks the update if:
        - The update fails for whatever reason (network issue, any error, etc)
        - The device is unplugged or there's a power interruption in the middle of an ongoing update

2. Selectively prune Docker images and volumes by filtering using labels.

3. Verify release signatures (should also be done in `install-box.sh` too).

---

### Before Merging

- [x] Update GitHub repository links to the official repository
- [x] Update Docker images to the official images
- [x] Enable pruning of all Docker images on successful update
- [x] Update configuration to run on mainnet 

---

### Related OTA Update PRs:

- https://github.com/getumbrel/umbrel-manager/pull/14
- https://github.com/getumbrel/umbrel-dashboard/pull/119

---
Will close #1 on merge.
